File: exercise3/library/src/main/java/com/example/library/base/BasePresenter.java
package com.example.library.base;

import android.support.annotation.NonNull;
import com.example.library.manager.RxManager;

/**
 * author : desperado
 * e-mail : foreverxiongtao@sina.com
 * date   : 2019/4/25 下午11:28
 * desc   : BasePresenter
 * version: 1.0
 */

public abstract class BasePresenter<M, V> {

    public M mModel;
    public V mView;
    protected RxManager mRxManager = new RxManager();

    /**
     * Returns the Model reference that the presenter wants to hold
     *
     * @return Model reference held by presenter
     */
    public abstract M getModel();

    /**
     * 绑定mModel和mView的引用
     *
     * @param m model
     * @param v view
     */
    public void attachMV(@NonNull M m, @NonNull V v) {
        this.mModel = m;
        this.mView = v;
        this.onStart();
    }

    /**
     * Untie mModel and mView
     */
    public void detachMV() {
        mRxManager.unSubscribe();
        mView = null;
        mModel = null;
    }

    /**
     * mView and mModel bindings are completed immediately
     * <p>
     * Implement the logic after the class implementation is complete, such as data initialization, interface initialization, update, etc.
     */
    public abstract void onStart();
}

File: app/src/main/java/com/example/exercise/presenter/BasePresenter.kt
package com.example.exercise.presenter

import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.LifecycleObserver
import android.arch.lifecycle.OnLifecycleEvent
import android.util.Log

/**
 *    author : desperado
 *    e-mail : foreverxiongtao@sina.com
 *    date   : 2019/5/7 下午4:58
 *    desc   :  LifeCycleObserver创建
 *    version: 1.0
 */
class BasePresenter : LifecycleObserver {
    private val TAG: String = BasePresenter::class.java.simpleName

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    fun onCreate() {
        Log.d(TAG, "onCreate")
    }


    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    fun onStart() {
        Log.i(TAG, "onStart()");
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    fun onResume() {
        Log.i(TAG, "onResume()");
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    fun onPause() {
        Log.i(TAG, "onPause()");
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    fun onStop() {
        Log.i(TAG, "onStop()");
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun onDestroy() {
        Log.d(TAG, "onDestroy");
    }
}